torch/testing/_internal/opinfo_helper.py

```diff
@@ -0,0 +1,145 @@
+                           all_types_and_half,
+                           all_types,
+                           complex_types,
+                           floating_and_complex_types,
+                           floating_types_and_half,
+                           floating_types,
+                           integral_types,
+                           floating_types_and,
+                           floating_and_complex_types_and,
+                           integral_types_and,
+                           all_types_and,
+                           )
+
+COMPLETE_DTYPES_DISPATCH = (
+    all_types,
+    all_types_and_complex,
+    all_types_and_half,
+    floating_types,
+    floating_and_complex_types,
+    floating_types_and_half,
+    integral_types,
+    complex_types,
+)
+
+EXTENSIBLE_DTYPE_DISPATCH = (
+    all_types_and_complex_and,
+    floating_types_and,
+    floating_and_complex_types_and,
+    integral_types_and,
+    all_types_and,
+)
+
+# Better way to acquire devices?
+DEVICES = ['cpu'] + (['cuda'] if TEST_CUDA else [])
+
+# Class to tag the dynamically generated types.
+
+
+class _dynamic_dispatch_dtypes(_dispatch_dtypes):
+    pass
+
+
+def get_supported_dtypes(op, sample_inputs_fn, device_type):
+    # Returns the supported dtypes for the given operator and device_type pair.
+    assert device_type in ['cpu', 'cuda']
+    if not TEST_CUDA and device_type == 'cuda':
+        warnings.warn("WARNING: CUDA is not available, information pertaining to CUDA dtypes could be wrong")
+        return _dynamic_dispatch_dtypes(())
+
+    supported_dtypes = set()
+    for dtype in all_types_and_complex_and(torch.bool, torch.bfloat16, torch.half):
+        samples = sample_inputs_fn(op, device_type, dtype, False)
+
+        # We assume the dtype is supported
+        # only if all samples pass for the given dtype.
+        supported = True
+        for sample in samples:
+            try:
+                op(sample.input, *sample.args, **sample.kwargs)
+            except RuntimeError as re:
+                if "not implemented for" not in str(re):
+                    raise re
+                else:
+                    # dtype is not supported
+                    supported = False
+
+        if supported:
+            supported_dtypes.add(dtype)
+
+    return _dynamic_dispatch_dtypes(supported_dtypes)
+
+
+def dtypes_dispatch_hint(dtypes):
+    return_type = collections.namedtuple('return_type', 'dispatch_fn dispatch_fn_str')
+
+    # CUDA is not available, dtypes will be empty.
+    if len(dtypes) == 0:
+        return return_type((), str(tuple()))
+
+    # Find the closest dispatcher from the set of available dispatchers
+
+    # Metric for finding the closest dispatcher
+    # Intersection Over Union
+    def iou(dispatch_dtypes):
+        return len(set(dtypes).intersection(dispatch_dtypes)) * 1.0 / len(set(dtypes).union(dispatch_dtypes))
+
+    dispatch_iou_score = {dispatch: iou(dispatch()) for dispatch in COMPLETE_DTYPES_DISPATCH}
```

Comment:
I still think we should simplify this for readability/maintainability with something like the following:

```
for dispatch in COMPLETE_DTYPES_DISPATCH:
  if dtypes == dispatch:
    # successful short-circuit goes here

closest_dispatch = None
closest_union_size = 0
for dispatch in EXTENSIBLE_DTYPE_DISPATCH:
  if not dispatch.issubset(dtypes):
    continue

  union_size = len(dtypes.union(dispatch))
  if union_size > closest_union_size:
    closest_union_size = union_size
    closest_dispatch = dispatch
```

Does that make sense? I think it just makes the logic a lot simpler to read

Reply:
(Also I think it's fewer lines of code)

Reply:
Have followed this structure. Just had to replace union_size with `iou` because union of `dispatch` and `dtypes` will always be `dtypes` as we make sure that `dispatch` is a subset of `dtypes`

Reply:
I added another inline comment about this. I agree with your analysis but derived a different conclusion.